src/linker.rs

```diff
@@ -507,24 +510,27 @@ where
         .create_module(c"linked_module")
         .ok_or(LinkerError::CreateModuleError)?;
 
-    // buffer used to perform file type detection
-    let mut buf = [0u8; 8];
-    for mut input in inputs {
+    for input in inputs {
         let path = match input {
             InputReader::File { path, .. } => path.into(),
             InputReader::Buffer { name, .. } => PathBuf::from(format!("in_memory::{}", name)),
         };
 
-        // determine whether the input is bitcode, ELF with embedded bitcode, an archive file
-        // or an invalid file
-        input
-            .read_exact(&mut buf)
+        let mut buf = BufReader::new(input);
+
+        // Peek at the buffer to determine file type
+        let preview = buf
```

Comment:
this is still an arbitrary size buffer (the size of the internal BufReader
buffer which is 4096 by default)

You don't need BufReader here. What you want to do is pass `input` to
detect_input_type instead of passing `&[u8]`